exporter/clickhouseexporter/exporter.go

```diff
@@ -47,20 +51,161 @@ func newExporter(logger *zap.Logger, cfg *Config) (*clickhouseExporter, error) {
+		keys = append(keys, formatKey(k))
+		values = append(values, v.AsString())
+		return true
+	})
+	return keys, values
+}
+
+func formatKey(k string) string {
+	return strings.ReplaceAll(k, ".", "_")
 }
 
+const (
+	// language=ClickHouse SQL
+	createLogsTableSQL = `
+CREATE TABLE IF NOT EXISTS logs_local (
+     Timestamp DateTime CODEC(Delta, ZSTD(1)),
+     TraceId String CODEC(ZSTD(1)),
+     SpanId String CODEC(ZSTD(1)),
+     TraceFlags UInt32,
+     SeverityText LowCardinality(String) CODEC(ZSTD(1)),
+     SeverityNumber Int32,
+     Body String CODEC(ZSTD(1)),
+     Attributes Nested
+     (
+         Key LowCardinality(String),
+         Value String
+     ) CODEC(ZSTD(1)),
+     Resource Nested
+     (
+         Key LowCardinality(String),
+         Value String
+     ) CODEC(ZSTD(1)),
+     INDEX idx_attr_keys Attributes.Key TYPE bloom_filter(0.01) GRANULARITY 64,
+     INDEX idx_res_keys Resource.Key TYPE bloom_filter(0.01) GRANULARITY 64
+) ENGINE MergeTree()
+%s
+PARTITION BY toDate(Timestamp)
+ORDER BY (toUnixTimestamp(Timestamp));
+`
+	// language=ClickHouse SQL
+	query = `INSERT INTO logs_local (
+                        Timestamp,
+                        TraceId,
+                        SpanId,
+                        TraceFlags,
+                        SeverityText,
+                        SeverityNumber,
+                        Body,
+                        Attributes.Key,
+                        Attributes.Value,
+                        Resource.Key, 
+                        Resource.Value
+                        ) VALUES (
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?,
+                                  ?
+                                  )`
+)
+
+var driverName = "clickhouse" // for testing
+
 // newClickhouseClient create a clickhouse client.
 func newClickhouseClient(cfg *Config) (*sql.DB, error) {
-	// TODO implement create clickhouse client
-	return nil, nil
+	// use empty database to create database
+	db, err := sql.Open(driverName, cfg.DSN)
+	if err != nil {
+		return nil, fmt.Errorf("sql.Open:%w", err)
+	}
+	// create table
+	query := fmt.Sprintf(createLogsTableSQL, "")
+	if cfg.TTLDays > 0 {
+		query = fmt.Sprintf(createLogsTableSQL,
+			fmt.Sprintf(`TTL Timestamp + INTERVAL %d DAY`, cfg.TTLDays))
+	}
+	if _, err := db.Exec(query); err != nil {
+		return nil, fmt.Errorf("exec create table sql: %w", err)
+	}
+	return db, nil
+}
+
+func doWithTx(_ context.Context, db *sql.DB, fn func(tx *sql.Tx) error) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return fmt.Errorf("db.Begin: %w", err)
+	}
+	defer func() {
+		_ = tx.Rollback()
```

Comment:
Does it mean that Rollback will be executed regardless of the error?

Reply:
No, it will return  ErrTxDone if tx is commit https://cs.opensource.google/go/go/+/master:src/database/sql/sql.go;l=2304?q=rollback&ss=go%2Fgo.
It just a common usage, like this example: https://pkg.go.dev/database/sql#example-Tx.Prepare